proposed/resource-limit.md

```diff
@@ -0,0 +1,559 @@
+
+Users will interact with this component in order to obtain decisions for rate or concurrency limits. This abstraction require explicit release sematics to accommodate non self-replenishing resources. This component encompasses the TryAcquire/AcquireAsync mechanics (i.e. check vs wait behaviours) and default implementations will be provided for select accounting method (fixed window, sliding window, token bucket, concurrency). The return type is a `Resource` which manages the lifecycle of the aquired resources.
+
+#### Resource limit implementations
+
+In addition to defining the abstractions, we would also like to ship some default implementations of common types of resources limiters. Details on how the functionality described here are achieved can be found in the [implementation section](#concrete-implementations).
+
+1. Fixed window rate limiter
+
+This is the simplest type of rate limiter. It counts the number of resources acquired in a time span. If the resource limit is reached, additional calls to `TryAcquire` will return `false` and additional calls to `AcquireAsync` will wait if queuing is enabled until the queue limit. The resource is refreshed when the current window has elapsed.
+
+2. Sliding window rate limiter
+
+This is another type of rate limiter. Counts for the number of resources acquired in a time segment is kept. A window is defined as multiple time segments and the resource is considered to be exhausted if the total across the segments in the window is exceeded. If the resource is exhausted, additional calls to `TryAcquire` will return `false` and additional calls to `AcquireAsync` will wait if queuing is enabled until the queue limit. The window is updated when the current time segment has elapsed.
+
+As an example consider the following example:
+
+```
+Settings: window time span = 3s, number of segments per window = 3, resource limit = 10
+
+Time:   |1s |2s |3s |4s |
+Count:  [3] [4] [3] [1]
+        | Window 1  |
+            | Window 2  |
+```
+
+In the above example, resource is exhaused in Window 1 since the total count is 10 and no additional resources can be acquired in the 3rd second. On the 4th second, Window 2 now has a count 8 and can accept additional resource acquisition requests.
+
+3. Token bucket rate limiter
+
+This rate limiter uses the [token bucket algorithm](https://en.wikipedia.org/wiki/Token_bucket). The functionality will be similar to the two implementation described above but will be configured differently.
+
+4. Concurrency limiter
+
+This limiter keeps track of the number of concurrent resources in use. Conceptually, it will function in a similar way to a Semaphore/SemaphoreSlim but adapted to the proposed API. Like the rate limits above, it will also keep a queue of pending acquisition requests if configured to do so.
+
+#### Resource limit adoption
+
+Two BCL types where adoption of the resource limiters are beneficial and relatively simple are:
+
+- System.Threading.Channels
```

Comment:
Is there a reason why we are prioritizing the `Channels` here? Is the ASP.NET part of the work going to take advantage of it?

Reply:
It's to help flesh out the design, not that we have an immediate need. Channels and Pipelines were the most obvious places to slot this into the BCL to make sure it could be used at this level in the stack (they basically do this in a very specific way today internally)

Reply:
FWIW, if we add a new feature without feedback loop, we typically get it wrong. If we do not have immediate use for Channels, I think it would be better to move it to future with the other BCL types.

I assume that the support in Pipelines will be used by ASP.NET. Is that right?

Reply:
Let me be clear, we already have 3 places where we are *concretely* going to be using it in ASP.NET Core, but we know the abstraction does not belong in ASP.NET Core. To make sure we didn't design something that can only be used in ASP.NET Core, we looked for places that do similar things in the BCL and @JunTaoLuo wrote prototypes to make sure the design could be applied to those types (and it works).

That is the purpose for listing Channels and pipelines.

Reply:
Is Pipelines support for resource limits going to be used by ASP.NET in .NET 6?

Reply:
I don't know if we need that in .NET 6 no. What does that change?

Reply:
It means that we won't have the feedback loop to ensure that it is done right.

Reply:
One option could be to replat the current limits on top of resource limiters.